op-chain-ops/cmd/op-migrate/main.go

```diff
@@ -0,0 +1,310 @@
+		currentHash = header.ParentHash
+	}
+
+	log.Info("Successfully cleaned old blocks")
+
+	// We're done messing around with the database, so we can now commit the changes to the DB.
+	// Note that this doesn't actually write the changes to disk.
+	// log.Info("Committing state DB")
+	// newRoot, err := db.Commit(true)
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	log.Info("Creating new Genesis block")
+	// Create the header for the Bedrock transition block.
+	cel2Header := &types.Header{
+		ParentHash:  common.Hash{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		UncleHash:   types.EmptyUncleHash,
+		Coinbase:    predeploys.SequencerFeeVaultAddr, // TODO
+		Root:        trieRoot,
+		TxHash:      types.EmptyRootHash,
+		ReceiptHash: types.EmptyRootHash,
+		Bloom:       types.Bloom{},
+		Difficulty:  common.Big0,
+		Number:      common.Big0,
+		GasLimit:    (uint64)(20_000_000),
+		GasUsed:     0,
+		Time:        uint64(12345),
+		Extra:       []byte("CeL2"),
+		MixDigest:   common.Hash{},
+		Nonce:       types.BlockNonce{},
+		BaseFee:     big.NewInt(params.InitialBaseFee),
+	}
+
+	// Create the Bedrock transition block from the header. Note that there are no transactions,
+	// uncle blocks, or receipts in the Bedrock transition block.
+	cel2Block := types.NewBlock(cel2Header, nil, nil, nil, trie.NewStackTrie(nil))
+
+	// We did it!
+	log.Info(
+		"Built Bedrock transition",
+		"hash", cel2Block.Hash(),
+		"root", cel2Block.Root(),
+		"number", cel2Block.NumberU64(),
+		"gas-used", cel2Block.GasUsed(),
+		"gas-limit", cel2Block.GasLimit(),
+	)
+
+	log.Info("Header", "header", cel2Header)
+	log.Info("Body", "Body", cel2Block)
+
+	// Create the result of the migration.
+	res := &MigrationResult{
+		TransitionHeight:    cel2Block.NumberU64(),
+		TransitionTimestamp: cel2Block.Time(),
+		TransitionBlockHash: cel2Block.Hash(),
+	}
+
+	// If we're not actually writing this to disk, then we're done.
+	if !commit {
+		log.Info("Dry run complete")
+		return res, nil
+	}
+
+	// Otherwise we need to write the changes to disk. First we commit the state changes.
+	// log.Info("Committing trie DB")
+	// if err := db.Database().TrieDB().Commit(newRoot, true); err != nil {
+	// 	return nil, err
+	// }
+
+	// Next we write the Cel2 genesis block to the database.
+	rawdb.WriteTd(ldb, cel2Block.Hash(), cel2Block.NumberU64(), cel2Block.Difficulty())
+	rawdb.WriteBlock(ldb, cel2Block)
+	rawdb.WriteReceipts(ldb, cel2Block.Hash(), cel2Block.NumberU64(), nil)
+	rawdb.WriteCanonicalHash(ldb, cel2Block.Hash(), cel2Block.NumberU64())
+	rawdb.WriteHeadBlockHash(ldb, cel2Block.Hash())
+	rawdb.WriteHeadFastBlockHash(ldb, cel2Block.Hash())
+	rawdb.WriteHeadHeaderHash(ldb, cel2Block.Hash())
+
+	// TODO
+	// Make the first CeL2 block a finalized block.
+	rawdb.WriteFinalizedBlockHash(ldb, cel2Block.Hash())
+
+	// Set the standard options.
+	// TODO: What about earlier hardforks
```

Comment:
Any number <= `cel2Block.Number()` would have the same effect for a hardfork, since the number is only checked in statements like `if current_block >= hardfork_block`, right?

Reply:
True, I was mostly wondering about hardfork like Berlin, which aren't explicitly set here. My understanding is that those have always been enabled on Optimism, but we might need to set them.